// prepack/lib/evaluators/ForStatement.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreatePerIterationEnvironment = CreatePerIterationEnvironment;

exports.default = function (ast, strictCode, env, realm, labelSet) {
  let { init, test, update, body } = ast;

  if (init && init.type === "VariableDeclaration") {
    if (init.kind === "var") {
      // for (var VariableDeclarationList; Expression; Expression) Statement
      // 1. Let varDcl be the result of evaluating VariableDeclarationList.
      let varDcl = env.evaluate(init, strictCode);

      // 2. ReturnIfAbrupt(varDcl).
      varDcl;

      // 3. Return ? ForBodyEvaluation(the first Expression, the second Expression, Statement, « », labelSet).
      return ForBodyEvaluation(realm, test, update, body, [], labelSet, strictCode);
    } else {
      // for (LexicalDeclaration Expression; Expression) Statement
      // 1. Let oldEnv be the running execution context's LexicalEnvironment.
      let oldEnv = env;

      // 2. Let loopEnv be NewDeclarativeEnvironment(oldEnv).
      let loopEnv = _singletons.Environment.NewDeclarativeEnvironment(realm, oldEnv);

      // 3. Let loopEnvRec be loopEnv's EnvironmentRecord.
      let loopEnvRec = loopEnv.environmentRecord;

      // 4. Let isConst be the result of performing IsConstantDeclaration of LexicalDeclaration.
      let isConst = init.kind === "const";

      // 5. Let boundNames be the BoundNames of LexicalDeclaration.
      let boundNames = _singletons.Environment.BoundNames(realm, init);

      // 6. For each element dn of boundNames do
      for (let dn of boundNames) {
        // a. If isConst is true, then
        if (isConst) {
          // i. Perform ! loopEnvRec.CreateImmutableBinding(dn, true).
          loopEnvRec.CreateImmutableBinding(dn, true);
        } else {
          // b. Else,
          // i. Perform ! loopEnvRec.CreateMutableBinding(dn, false).
          loopEnvRec.CreateMutableBinding(dn, false);
        }
      }

      // 7. Set the running execution context's LexicalEnvironment to loopEnv.
      realm.getRunningContext().lexicalEnvironment = loopEnv;

      // 8. Let forDcl be the result of evaluating LexicalDeclaration.
      let forDcl = loopEnv.evaluateCompletion(init, strictCode);

      // 9. If forDcl is an abrupt completion, then
      if (forDcl instanceof _completions.AbruptCompletion) {
        // a. Set the running execution context's LexicalEnvironment to oldEnv.
        let currentEnv = realm.getRunningContext().lexicalEnvironment;
        realm.onDestroyScope(currentEnv);
        if (currentEnv !== loopEnv) (0, _invariant2.default)(loopEnv.destroyed);
        realm.getRunningContext().lexicalEnvironment = oldEnv;

        // b. Return Completion(forDcl).
        throw forDcl;
      }

      // 10. If isConst is false, let perIterationLets be boundNames; otherwise let perIterationLets be « ».
      let perIterationLets = !isConst ? boundNames : [];

      let bodyResult;
      try {
        // 11. Let bodyResult be ForBodyEvaluation(the first Expression, the second Expression, Statement, perIterationLets, labelSet).
        bodyResult = ForBodyEvaluation(realm, test, update, body, perIterationLets, labelSet, strictCode);
      } finally {
        // 12. Set the running execution context's LexicalEnvironment to oldEnv.
        let currentEnv = realm.getRunningContext().lexicalEnvironment;
        realm.onDestroyScope(currentEnv);
        if (currentEnv !== loopEnv) (0, _invariant2.default)(loopEnv.destroyed);
        realm.getRunningContext().lexicalEnvironment = oldEnv;
      }
      // 13. Return Completion(bodyResult).
      return bodyResult;
    }
  } else {
    // for (Expression; Expression; Expression) Statement
    // 1. If the first Expression is present, then
    if (init) {
      // a. Let exprRef be the result of evaluating the first Expression.
      let exprRef = env.evaluate(init, strictCode);

      // b. Perform ? GetValue(exprRef).
      _singletons.Environment.GetValue(realm, exprRef);
    }

    // 2. Return ? ForBodyEvaluation(the second Expression, the third Expression, Statement, « », labelSet).
    return ForBodyEvaluation(realm, test, update, body, [], labelSet, strictCode);
  }
};

var _index = require("../values/index.js");

var _completions = require("../completions.js");

var _errors = require("../errors.js");

var _index2 = require("../methods/index.js");

var _ForOfStatement = require("./ForOfStatement.js");

var _singletons = require("../singletons.js");

var _invariant = require("../invariant.js");

var _invariant2 = _interopRequireDefault(_invariant);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// ECMA262 13.7.4.9
function CreatePerIterationEnvironment(realm, perIterationBindings) {
  // 1. If perIterationBindings has any elements, then
  if (perIterationBindings.length > 0) {
    // a. Let lastIterationEnv be the running execution context's LexicalEnvironment.
    let lastIterationEnv = realm.getRunningContext().lexicalEnvironment;
    // b. Let lastIterationEnvRec be lastIterationEnv's EnvironmentRecord.
    let lastIterationEnvRec = lastIterationEnv.environmentRecord;
    // c. Let outer be lastIterationEnv's outer environment reference.
    let outer = lastIterationEnv.parent;
    // d. Assert: outer is not null.
    (0, _invariant2.default)(outer !== null);
    // e. Let thisIterationEnv be NewDeclarativeEnvironment(outer).
    let thisIterationEnv = _singletons.Environment.NewDeclarativeEnvironment(realm, outer);
    // f. Let thisIterationEnvRec be thisIterationEnv's EnvironmentRecord.
    realm.onDestroyScope(lastIterationEnv);
    let thisIterationEnvRec = thisIterationEnv.environmentRecord;
    // g. For each element bn of perIterationBindings do,
    for (let bn of perIterationBindings) {
      // i. Perform ! thisIterationEnvRec.CreateMutableBinding(bn, false).
      thisIterationEnvRec.CreateMutableBinding(bn, false);
      // ii. Let lastValue be ? lastIterationEnvRec.GetBindingValue(bn, true).
      let lastValue = lastIterationEnvRec.GetBindingValue(bn, true);
      // iii.Perform thisIterationEnvRec.InitializeBinding(bn, lastValue).
      thisIterationEnvRec.InitializeBinding(bn, lastValue);
    }
    // h. Set the running execution context's LexicalEnvironment to thisIterationEnv.
    realm.getRunningContext().lexicalEnvironment = thisIterationEnv;
  }
  // 2. Return undefined.
  return realm.intrinsics.undefined;
}

// ECMA262 13.7.4.8
/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

function ForBodyEvaluation(realm, test, increment, stmt, perIterationBindings, labelSet, strictCode) {
  // 1. Let V be undefined.
  let V = realm.intrinsics.undefined;

  // 2. Perform ? CreatePerIterationEnvironment(perIterationBindings).
  CreatePerIterationEnvironment(realm, perIterationBindings);
  let env = realm.getRunningContext().lexicalEnvironment;

  // 3. Repeat
  while (true) {
    // a. If test is not [empty], then
    if (test) {
      // i. Let testRef be the result of evaluating test.
      let testRef = env.evaluate(test, strictCode);

      // ii. Let testValue be ? GetValue(testRef).
      let testValue = _singletons.Environment.GetValue(realm, testRef);

      // iii. If ToBoolean(testValue) is false, return NormalCompletion(V).
      if (!_singletons.To.ToBooleanPartial(realm, testValue)) {
        // joinAllLoopExits does not handle labeled break/continue, so only use it when doing AI
        if (realm.useAbstractInterpretation) return joinAllLoopExits(V);
        return V;
      }
    }

    // b. Let result be the result of evaluating stmt.
    let result = env.evaluateCompletion(stmt, strictCode);
    (0, _invariant2.default)(result instanceof _index.Value || result instanceof _completions.AbruptCompletion);

    // c. If LoopContinues(result, labelSet) is false, return Completion(UpdateEmpty(result, V)).
    if (!(0, _ForOfStatement.LoopContinues)(realm, result, labelSet)) {
      (0, _invariant2.default)(result instanceof _completions.AbruptCompletion);
      // joinAllLoopExits does not handle labeled break/continue, so only use it when doing AI
      if (realm.useAbstractInterpretation) {
        result = (0, _index2.UpdateEmpty)(realm, result, V);
        (0, _invariant2.default)(result instanceof _completions.AbruptCompletion);
        return joinAllLoopExits(result);
      }
      // ECMA262 13.1.7
      if (result instanceof _completions.BreakCompletion) {
        if (!result.target) return (0, _index2.UpdateEmpty)(realm, result, V).value;
      }
      throw (0, _index2.UpdateEmpty)(realm, result, V);
    } else if (realm.useAbstractInterpretation) {
      // This is a join point for conditional continue completions lurking in realm.savedCompletion
      if (containsContinueCompletion(realm.savedCompletion)) {
        result = joinAllLoopContinues(result);
      }
    }

    // d. If result.[[Value]] is not empty, let V be result.[[Value]].
    let resultValue = (0, _ForOfStatement.InternalGetResultValue)(realm, result);
    if (!(resultValue instanceof _index.EmptyValue)) V = resultValue;

    // e. Perform ? CreatePerIterationEnvironment(perIterationBindings).
    CreatePerIterationEnvironment(realm, perIterationBindings);
    env = realm.getRunningContext().lexicalEnvironment;

    // f. If increment is not [empty], then
    if (increment) {
      // i. Let incRef be the result of evaluating increment.
      let incRef = env.evaluate(increment, strictCode);

      // ii. Perform ? GetValue(incRef).
      _singletons.Environment.GetValue(realm, incRef);
    }
  }
  (0, _invariant2.default)(false);

  function failIfContainsBreakOrContinueCompletionWithNonLocalTarget(c) {
    if (c === undefined) return;
    if (c instanceof _completions.ContinueCompletion || c instanceof _completions.BreakCompletion) {
      if (!c.target) return;
      if (labelSet && labelSet.indexOf(c.target) >= 0) {
        c.target = null;
        return;
      }
      let diagnostic = new _errors.CompilerDiagnostic("break or continue with target cannot be guarded by abstract condition", c.location, "PP0034", "FatalError");
      realm.handleError(diagnostic);
      throw new _errors.FatalError();
    }
    if (c instanceof _completions.PossiblyNormalCompletion || c instanceof _completions.JoinedAbruptCompletions) {
      failIfContainsBreakOrContinueCompletionWithNonLocalTarget(c.consequent);
      failIfContainsBreakOrContinueCompletionWithNonLocalTarget(c.alternate);
    }
  }

  function containsContinueCompletion(c) {
    if (c === undefined) return false;
    if (c instanceof _completions.ContinueCompletion) {
      if (!c.target) return true;
      if (labelSet && labelSet.indexOf(c.target) >= 0) {
        c.target = null;
        return true;
      }
      return false;
    }
    if (c instanceof _completions.PossiblyNormalCompletion || c instanceof _completions.JoinedAbruptCompletions) return containsContinueCompletion(c.consequent) || containsContinueCompletion(c.alternate);
    return false;
  }

  function joinAllLoopContinues(valueOrCompletionAtLoopContinuePoint) {
    // We are about start the next loop iteration and this presents a join point where all non loop breaking abrupt
    // control flows converge into a single flow using their joined effects as the new state.
    failIfContainsBreakOrContinueCompletionWithNonLocalTarget(realm.savedCompletion);

    // Incorporate the savedCompletion (we should only get called if there is one).
    (0, _invariant2.default)(realm.savedCompletion !== undefined);
    if (valueOrCompletionAtLoopContinuePoint instanceof _index.Value) valueOrCompletionAtLoopContinuePoint = new _completions.ContinueCompletion(valueOrCompletionAtLoopContinuePoint);
    let abruptCompletion = _singletons.Functions.incorporateSavedCompletion(realm, valueOrCompletionAtLoopContinuePoint);
    (0, _invariant2.default)(abruptCompletion instanceof _completions.AbruptCompletion);

    // If there is now a single completion, we don't need to join
    if (!(abruptCompletion instanceof _completions.JoinedAbruptCompletions)) return abruptCompletion;
    (0, _invariant2.default)(containsContinueCompletion(abruptCompletion));

    // Apply the joined effects of continue completions to the current state since these now join the normal path
    let joinedContinueEffects = _singletons.Join.extractAndJoinCompletionsOfType(_completions.ContinueCompletion, realm, abruptCompletion);
    realm.applyEffects(joinedContinueEffects);
    let c = joinedContinueEffects.result;
    (0, _invariant2.default)(c instanceof _completions.ContinueCompletion);

    // We now make a PossiblyNormalCompletion out of abruptCompletion.
    // extractAndJoinCompletionsOfType helped with this by cheating and turning all of its nested completions
    // that contain continue completions into PossiblyNormalCompletions.
    let remainingCompletions = abruptCompletion.transferChildrenToPossiblyNormalCompletion();

    // At this stage there can still be other kinds of abrupt completions left inside abruptCompletion. If not just return.
    let stillAbrupt = remainingCompletions.containsCompletion(_completions.BreakCompletion) || remainingCompletions.containsCompletion(_completions.ReturnCompletion) || remainingCompletions.containsCompletion(_completions.ThrowCompletion);
    if (!stillAbrupt) return c;

    // Stash the remaining completions in the realm start tracking the effects that need to be appended
    // to the normal branch at the next join point.
    realm.savedCompletion = remainingCompletions;
    realm.captureEffects(remainingCompletions); // so that we can join the normal path wtih them later on
    return c;
  }

  function joinAllLoopExits(valueOrCompletionAtUnconditionalExit) {
    // We are about the leave this loop and this presents a join point where all loop breaking control flows
    // converge into a single flow using their joined effects as the new state.
    failIfContainsBreakOrContinueCompletionWithNonLocalTarget(realm.savedCompletion);

    // Incorporate the savedCompletion if there is one.
    if (valueOrCompletionAtUnconditionalExit instanceof _index.Value) valueOrCompletionAtUnconditionalExit = new _completions.BreakCompletion(valueOrCompletionAtUnconditionalExit);
    let abruptCompletion = _singletons.Functions.incorporateSavedCompletion(realm, valueOrCompletionAtUnconditionalExit);
    (0, _invariant2.default)(abruptCompletion instanceof _completions.AbruptCompletion);

    // If there is now a single completion, we don't need to join
    if (abruptCompletion instanceof _completions.BreakCompletion) return (0, _index2.UpdateEmpty)(realm, abruptCompletion, V).value;
    if (!(abruptCompletion instanceof _completions.JoinedAbruptCompletions)) throw abruptCompletion;

    // If there are no breaks, we don't need to join
    if (!abruptCompletion.containsCompletion(_completions.BreakCompletion)) throw abruptCompletion;

    // Apply the joined effects of break completions to the current state since these now join the normal path
    let joinedBreakEffects = _singletons.Join.extractAndJoinCompletionsOfType(_completions.BreakCompletion, realm, abruptCompletion);
    realm.applyEffects(joinedBreakEffects);
    let c = joinedBreakEffects.result;
    (0, _invariant2.default)(c instanceof _completions.BreakCompletion);

    // We now make a PossiblyNormalCompletion out of abruptCompletion.
    // extractAndJoinCompletionsOfType helped with this by cheating and turning all of its nested completions
    // that contain continue completions into PossiblyNormalCompletions.
    let remainingCompletions = abruptCompletion.transferChildrenToPossiblyNormalCompletion();

    // At this stage there can still be other kinds of abrupt completions left inside abruptCompletion. If not just return.
    let stillAbrupt = remainingCompletions.containsCompletion(_completions.ReturnCompletion) || remainingCompletions.containsCompletion(_completions.ThrowCompletion);
    if (!stillAbrupt) return (0, _index2.UpdateEmpty)(realm, c, V).value;

    // Stash the remaining completions in the realm start tracking the effects that need to be appended
    // to the normal branch at the next join point.
    realm.savedCompletion = remainingCompletions;
    realm.captureEffects(remainingCompletions); // so that we can join the normal path wtih them later on

    // ECMA262 13.1.7
    return (0, _index2.UpdateEmpty)(realm, c, V).value;
  }
}

// ECMA262 13.7.4.7
